### PR TITLE
fix: FPs with authentication

### PR DESCRIPTION
**Pull request by EsadCetiner**
Fixes various false positives with authentication such as logging in in webauthn, cookies and resetting passwords

### CODE REVIEW

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 1]
```diff
@@ -739,15 +792,38 @@ SecRule REQUEST_FILENAME "@endsWith /login" \
             ctl:ruleRemoveTargetByTag=OWASP_CRS;ARGS:pass1-text,\
             ctl:ruleRemoveTargetByTag=OWASP_CRS;ARGS:pass2"
 
-# Logout token
-SecRule REQUEST_FILENAME "@endsWith /logout" \
-    "id:9508420,\
+# Password reset used in newer version of Nextcloud
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:795]:
```suggestion
# Password reset used in newer versions of Nextcloud
```

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 2]
```diff
@@ -739,15 +792,38 @@ SecRule REQUEST_FILENAME "@endsWith /login" \
...
-    ctl:ruleRemoveTargetById=941120;ARGS:requesttoken"
+    ctl:ruleRemoveTargetByTag=OWASP_CRS;ARGS:json.password"
+
+# requesttoken is used when logging in or when authenticating via a public share
+# Matches:
+# /login
+# /s/share-id/authenticate/showShare
+# /login/v2/grant
+# /logout
+SecRule REQUEST_FILENAME "@rx /(?:login(?:/v[0-9\.]+/grant)?|s/[^/]+/authenticate/showShare|logout)$" \
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:811]: Should also match `/login`, to support original rule 9508400.

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 3]
```diff
@@ -739,15 +792,38 @@ SecRule REQUEST_FILENAME "@endsWith /login" \
...
+    SecRule ARGS:requesttoken "@rx ^[a-zA-Z0-9+=:/]+$" \
+        "t:none,\
+        ctl:ruleRemoveTargetById=920273;ARGS:requesttoken,\
+        ctl:ruleRemoveTargetById=932236;ARGS:requesttoken,\
+        ctl:ruleRemoveTargetById=941100;ARGS:requesttoken,\
+        ctl:ruleRemoveTargetById=941120;ARGS:requesttoken,\
+        ctl:ruleRemoveTargetById=942432;ARGS:requesttoken,\
+        ctl:ruleRemoveTargetById=942450;ARGS:requesttoken"
 
 # Logging in with webauthn
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:828]: This looks like a duplicate of the new rule 9508402.

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 2 above)
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:811]: This rule is complex. It should have tests.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            headers:
+              Host: localhost
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: POST
+            uri: /login
+            data: "redirect_url=/login/v2/grant?user=Esad%20Cetiner&stateToken=sample-token"
+          output:
+            no_log_contains: |
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml:55]: Careful! This is a regular expression, not a list.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            headers:
+              Host: localhost
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: POST
+            uri: /index.php/login
+            data: "redirect_url=/login/v2/grant?user=Esad%20Cetiner&stateToken=sample-token"
+          output:
+            no_log_contains: |
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml:77]: Careful! This is a regular expression, not a list.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml [hunk 6]
```diff
@@ -0,0 +1,83 @@
+---
+meta:
+  author: "Esad Cetiner"
+  description: "Nextcloud Rule Exclusions Plugin"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml:4]:
```suggestion
  description: "FPs during login"
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508401.yaml [hunk 7]
```diff
@@ -0,0 +1,71 @@
+---
+meta:
+  author: "Esad Cetiner"
+  description: "Nextcloud Rule Exclusions Plugin"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508401.yaml:4]:
```suggestion
  description: "FPs during reauthentication / verification"
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml [hunk 8]
```diff
@@ -0,0 +1,57 @@
+---
+meta:
+  author: "Esad Cetiner"
+  description: "Nextcloud Rule Exclusions Plugin"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml:4]:
```suggestion
  description: "FPs during login with FIDO2"
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+          input:
+            dest_addr: 127.0.0.1
+            headers:
+              Host: localhost
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: POST
+            uri: /login/webauthn/finish
+            data: "json.data={\"id\":\"RanDom__02Data\",\"type\":\"public-key\",\"rawId\":\"FDADG34//fsd88DF=\",\"response\":{\"authenticatorData\":\"3/fds+dsf79dDAF/SQ+sdfaf89sDFS==\",\"clientDataJSON\":\"7sdfybsdfbyubYUFDBVASUBYFASD7687gdsfb==\",\"signature\":\"dsyuifabHBJDAF989+sfddsfDFKJ6678+fsd676JKG+S8dg=\",\"userHandle\":\"random=\"}}"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml:21]: Use a scalar here, then you don't need to escape the double quotes:
```yaml
data: |
  json.data=...
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+          output:
+            no_log_contains: |
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml:23]: Careful! This is a regular expression, not a list.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+          input:
+            dest_addr: 127.0.0.1
+            headers:
+              Host: localhost
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: POST
+            uri: /index.php/login/webauthn/finish
+            data: "json.data={\"id\":\"RanDom__02Data\",\"type\":\"public-key\",\"rawId\":\"FDADG34//fsd88DF=\",\"response\":{\"authenticatorData\":\"3/fds+dsf79dDAF/SQ+sdfaf89sDFS==\",\"clientDataJSON\":\"7sdfybsdfbyubYUFDBVASUBYFASD7687gdsfb==\",\"signature\":\"dsyuifabHBJDAF989+sfddsfDFKJ6678+fsd676JKG+S8dg=\",\"userHandle\":\"random=\"}}"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml:46]: Use a scalar here, then you don't need to escape the double quotes:
```yaml
data: |
  json.data=...
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+          output:
+            no_log_contains: |
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508402.yaml:48]: Careful! This is a regular expression, not a list.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508403.yaml [hunk 13]
```diff
@@ -0,0 +1,71 @@
+---
+meta:
+  author: "Esad Cetiner"
+  description: "Nextcloud Rule Exclusions Plugin"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508403.yaml:4]:
```suggestion
  description: "FPs during login from desktop app"
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508411.yaml [hunk 14]
```diff
@@ -0,0 +1,39 @@
+---
+meta:
+  author: "Esad Cetiner"
+  description: "Nextcloud Rule Exclusions Plugin"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508411.yaml:4]:
```suggestion
  description: "FPs during password reset"
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml [hunk 15]
```diff
@@ -0,0 +1,57 @@
+---
+meta:
+  author: "Esad Cetiner"
+  description: "Nextcloud Rule Exclusions Plugin"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml:4]:
```suggestion
  description: "FPs when registering webauthn devices"
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+          input:
+            dest_addr: 127.0.0.1
+            headers:
+              Host: localhost
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: POST
+            uri: /settings/api/personal/webauthn/registration
+            data: "json.data={\"id\":\"RanDom__02Data\",\"type\":\"public-key\",\"rawId\":\"FDADG34//fsd88DF=\",\"response\":{\"authenticatorData\":\"3/fds+dsf79dDAF/SQ+sdfaf89sDFS==\",\"clientDataJSON\":\"7sdfybsdfbyubYUFDBVASUBYFASD7687gdsfb==\",\"signature\":\"dsyuifabHBJDAF989+sfddsfDFKJ6678+fsd676JKG+S8dg=\",\"userHandle\":\"random=\"}}"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml:21]: Use a scalar here, then you don't need to escape the double quotes:
```yaml
data: |
  json.data=...
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+          output:
+            no_log_contains: |
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml:23]: Careful! This is a regular expression, not a list.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+          input:
+            dest_addr: 127.0.0.1
+            headers:
+              Host: localhost
+              User-Agent: OWASP ModSecurity Core Rule Set
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: POST
+            uri: /settings/api/personal/webauthn/registration
+            data: "json.data={\"id\":\"RanDom__02Data\",\"type\":\"public-key\",\"rawId\":\"FDADG34//fsd88DF=\",\"response\":{\"authenticatorData\":\"3/fds+dsf79dDAF/SQ+sdfaf89sDFS==\",\"clientDataJSON\":\"7sdfybsdfbyubYUFDBVASUBYFASD7687gdsfb==\",\"signature\":\"dsyuifabHBJDAF989+sfddsfDFKJ6678+fsd676JKG+S8dg=\",\"userHandle\":\"random=\"}}"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml:46]: Use a scalar here, then you don't need to escape the double quotes:
```yaml
data: |
  json.data=...
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+          output:
+            no_log_contains: |
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508624.yaml:48]: Careful! This is a regular expression, not a list.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml [hunk 20]
```diff
@@ -20,84 +20,10 @@ tests:
...
-              Host: localhost
-              User-Agent: OWASP ModSecurity Core Rule Set
-              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
-            port: 80
-            method: POST
-            uri: /ocs/v2.php/references/resolve
-            data: "reference=https://www.jaycar.com.au/raspberry-pi-3-switchmode-power-supply-5-1v-2-5a-with-usb-micro-b/p/MP3536?pos=1&queryId=0a9ebf34f466c42c586d9561f4ac20d2&sort=relevance&searchText=2.5a%205v%20micro%20usb"
-          output:
-            no_log_contains: id "942432"
+            no_log_contains: |
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: Careful! This is a regular expression, not a list.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml [hunk 21]
```diff
@@ -20,84 +20,10 @@ tests:
             uri: /ocs/v2.php/references/resolve
             data: "reference=https://www.jaycar.com.au/raspberry-pi-3-switchmode-power-supply-5-1v-2-5a-with-usb-micro-b/p/MP3536?pos=1&queryId=0a9ebf34f466c42c586d9561f4ac20d2&sort=relevance&searchText=2.5a%205v%20micro%20usb"
           output:
-            no_log_contains: id "920230"
-  - test_title: 9508880-2
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:24]: Why did you remove these tests?

**Review by theseion — CHANGES_REQUESTED** [review]

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 2 above)
Comment (EsadCetiner) [reply plugins/nextcloud-rule-exclusions-before.conf:811]: it should match /login, does it not? 

or is the regex not clear enough, should I change to this?
```suggestion
SecRule REQUEST_FILENAME "@rx /(?:login|login/v[0-9\.]+/grant|s/[^/]+/authenticate/showShare|logout)$" \
```

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 3 above)
Comment (EsadCetiner) [reply plugins/nextcloud-rule-exclusions-before.conf:828]: it's similar, but it checks the response
I'll add a comment to make what it does clearer

File: tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml (hunk 21 above)
Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:24]: they all do the same thing, except they check for different rules. imo it looks cleaner this way since the test is testing for all the rules the payload triggers, instead of creating multiple tests for different rules.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml (hunk 20 above)
Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: so what should I do here then?
like I mentioned in your other review, I didn't want to create multiple tests, testing the same payload just to test different rules.

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 2 above)
Comment (EsadCetiner) [reply plugins/nextcloud-rule-exclusions-before.conf:811]: I don't have or know all the payloads for all of the rules triggered, so there won't be complete coverage.

Comment (theseion) [reply plugins/nextcloud-rule-exclusions-before.conf:811]: You're right. I missed the optional part there. Leave the expression as you wrote it but make sure to add `/login` to the comment above the rule.

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 3 above)
Comment (theseion) [reply plugins/nextcloud-rule-exclusions-before.conf:828]: Ah, ok.

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 2 above)
Comment (EsadCetiner) [reply plugins/nextcloud-rule-exclusions-before.conf:811]: that is also there 😄

File: tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml (hunk 20 above)
Comment (theseion) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: For example:
```yaml
            no_log_contains: |
              id 942(?:432|...)
```

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 2 above)
Comment (theseion) [reply plugins/nextcloud-rule-exclusions-before.conf:811]: No issue. But I think you should at least test that the regular expression matches the paths you listed in the comment.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml (hunk 20 above)
Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: gotcha, but that looks a bit worse in terms of readability, is this a performance change?

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 2 above)
Comment (theseion) [reply plugins/nextcloud-rule-exclusions-before.conf:811]: Lol. Sorry, must have been stricken with blindness when I did the review 😅

Comment (EsadCetiner) [reply plugins/nextcloud-rule-exclusions-before.conf:811]: happens to the best of us

File: tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml (hunk 20 above)
Comment (theseion) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: Maybe not as readable, but, as you wrote about the tests you removed, more readable than duplicated tests with different `no_log_contains:`.

> is this a performance change?

I don't understand the question. Are you asking whether it's a new thing that `no_log_contains` is a regular expression?

Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: no, I'm just trying to understand the reasoning for change

Comment (theseion) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: Not sure, what you are referring to. `log_contains` and `no_log_contains` have been regular expressions for a long time (maybe always, not sure).

Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: I just mean why I should use a non capturing group to reduce the amount of rule ID entries like in your example?
I think it's perfectly fine to list all of the rule IDs as is

Comment (theseion) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: Oh, I see :D. Yes, feel free to drop the group, that was just an example.

Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: so I can leave them as is then as just a list?

Comment (theseion) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: No, you have to at least transform the list into `id (?:942432|942...)`.  What you currently have there would amount to `id 942432\nid 942...`. So you'd be looking for the _entire string_.

Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: ok... everything should be good then

Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: ooh, you just want all of the rule IDs to be on the same line, you don't care if I use something like non-capturing groups or character sets, or do you want both?

Comment (theseion) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: You can write the regex however you like :)

Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: is this better?

Comment (theseion) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:23]: Almost, but now you have extra spaces that are part of your match expression.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml [hunk 22]
```diff
@@ -20,84 +20,5 @@ tests:
...
-              User-Agent: OWASP ModSecurity Core Rule Set
-              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
-            port: 80
-            method: POST
-            uri: /ocs/v2.php/references/resolve
-            data: "reference=https://www.jaycar.com.au/raspberry-pi-3-switchmode-power-supply-5-1v-2-5a-with-usb-micro-b/p/MP3536?pos=1&queryId=0a9ebf34f466c42c586d9561f4ac20d2&sort=relevance&searchText=2.5a%205v%20micro%20usb"
-          output:
-            no_log_contains: id "942432"
+            no_log_contains: |
+              id "920230" | id "93(?:1130|2200)" | id "94243[012]"
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml:24]:
```suggestion
              id "920230"|id "93(?:1130|2200)"|id "94243[012]"
```

Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508880.yaml]: done

**Review by theseion — CHANGES_REQUESTED** [review]

**Review by theseion — APPROVED** [review]
